### ISSUE TITLE
[Question] - domain inheritance with pattern matching

### ISSUE THREAD

**Issue by peterzernia** [casbin/casbin#1157]
I want to have a RBAC model where there are multiple domains and subdomains. The roles across all domains can be the same with access to the same resources, but they should only be allowed access within a specific subdomain/domain

**Your model:**

```ini
r = sub, dom, obj, act

[policy_definition]
p = sub, dom, obj, act

[role_definition]
g = _, _, _
g2 = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = (g(r.sub, p.sub, r.dom) || g2(r.dom, p.dom)) && \
r.obj == p.obj && \
r.act == p.act
```

**Your policy:**

```
p, user, *, data, read
p, user, *, data, write
p, admin, *, data, read
p, admin, *, data, write

g, alice, user, subdomain1
g, bob, admin, domain1

g2, subdomain1, domain1
g2, subdomain2, domain1

```

**Your request(s):**

```
alice, subdomain1, data, read --> true (expected true)
bob, domain1, data, read --> true (expected true)
bob, subdomain1, data, read --> false( expected true)
bob, subdomain2, data, read --> false (expected true
```

In this case, bob is an admin of domain1, and subdomain1 and subdomain2 are part of domain1, which he should automatically inherit access to. How can I get the `|| g2(r.dom, p.dom)` part of to use `keyMatch`. It is changing subdomain1 in the request to domain1, but then matching domain1 and *, which only match through keyMatch

**Comment by peterzernia** [comment]
adding `e.AddNamedMatchingFunc("g2","KeyMatch",util.KeyMatch)` matches the * to every domain meaning
```
bob, subdomain3, data, read ---> true (expected false)
```
this becomes true when subdomain3 is not inherited

**Comment by hsluoyz** [comment]
@maocaicai

**Comment by peterzernia** [comment]
is there any update on this? Or is it just not possible?

**Comment by yyy1000** [comment]
Hi,
I'm afraid the "*" in the policy for your model can't work. 
You could try to change it to 'domain1' like below
```
 p, user, domain1, data, read
p, admin, domain1, data, read

g, alice, user, subdomain1
g, bob, admin, domain1

g2, subdomain1, domain1
g2, subdomain2, domain1
```

**Comment by ZhengLin-Li** [comment]
I agree with this comment https://github.com/casbin/casbin/issues/1157#issuecomment-1373628429 , maybe you could try just list them all out.

**Comment by yyy1000** [comment]
> Thanks. A lot of the roles will be the same across domains. There is no other way to abstract a role?
Hi, maybe you can try this? I tested and get the same result but I'm not sure whether it works for you.
```
p, user, *, data, read
p, admin, *, data, read

g, alice, user, subdomain1
g, bob, admin, domain1

g2, domain1, *
g2, subdomain1, domain1
g2, subdomain2, domain1
```

We do plan to support "*" in your case, but it may take some time. :)

**Comment by yyy1000** [comment]
Oh
I just get clear.

In your case, if you want to use "*" to represent "all the domains" (with keyMatch function), this will include all domains even if this domain doesn't appear, like "subdomain3", so pointing out the domain like the two ways above is a way.
And another thing that needs to pay attention to is the "data". In most cases,  you may need different data in different domains, so listing domains seems proper.
I'm working on this, feel free to ask me.

**Comment by peterzernia** [comment]
> > Thanks. A lot of the roles will be the same across domains. There is no other way to abstract a role?
> 
> Hi, maybe you can try this? I tested and get the same result but I'm not sure whether it works for you.
> 
> ```
> p, user, *, data, read
> p, admin, *, data, read
> 
> g, alice, user, subdomain1
> g, bob, admin, domain1
> 
> g2, domain1, *
> g2, subdomain1, domain1
> g2, subdomain2, domain1
> ```
> 
> It seems that there's no other simple and clear way to do it. We could try to support "*" in your case, but it may take some time. :)

Thanks for looking into this :). This almost works, but when I add another domain e.g. 
```
g2, domain3, *
g2, subdomain3, domain3
```
then
```
bob, subdomain3, data, read ---> true (expected false)
```
since bob was assigned to domain1, bob shouldn't have access to the top level domain3, but is getting it through the *.


in the policy, I want to use the `*` to represent that the permission (and role that inherits it) exist across all domains, but should only be allowed access in their domain. I will also have some permissions and roles that are very domain specific e.g.

```
p, domain3_user, domain3, domain3_data, read

g, lisa, domain3_user, domain3
```
